File: kreuzberg/packages/java/src/main/java/dev/kreuzberg/config/AccelerationConfig.java
package dev.kreuzberg.config;

import java.util.HashMap;
import java.util.Map;

/**
 * Configuration for ONNX Runtime execution provider selection.
 *
 * @since 4.5.0
 */
public final class AccelerationConfig {
	private final String provider;
	private final int deviceId;

	private AccelerationConfig(Builder builder) {
		this.provider = builder.provider;
		this.deviceId = builder.deviceId;
	}

	public static Builder builder() {
		return new Builder();
	}

	/**
	 * Get the execution provider name.
	 *
	 * @return the provider (auto, cpu, coreml, cuda, tensorrt)
	 */
	public String getProvider() {
		return provider;
	}

	/**
	 * Get the GPU device ID.
	 *
	 * @return the device ID (for CUDA/TensorRT)
	 */
	public int getDeviceId() {
		return deviceId;
	}

	public Map<String, Object> toMap() {
		Map<String, Object> map = new HashMap<>();
		if (provider != null && !provider.isEmpty()) {
			map.put("provider", provider);
		}
		if (deviceId != 0) {
			map.put("device_id", deviceId);
		}
		return map;
	}

	public static final class Builder {
		private String provider = "auto";
		private int deviceId = 0;

		private Builder() {
		}

		/**
		 * Set the execution provider.
		 *
		 * @param provider
		 *            the provider name (auto, cpu, coreml, cuda, tensorrt)
		 * @return this builder for chaining
		 */
		public Builder provider(String provider) {
			this.provider = provider;
			return this;
		}

		/**
		 * Set the GPU device ID.
		 *
		 * @param deviceId
		 *            the device ID for CUDA/TensorRT
		 * @return this builder for chaining
		 */
		public Builder deviceId(int deviceId) {
			this.deviceId = deviceId;
			return this;
		}

		public AccelerationConfig build() {
			return new AccelerationConfig(this);
		}
	}

	static AccelerationConfig fromMap(Map<String, Object> map) {
		if (map == null) {
			return null;
		}
		Builder builder = builder();
		if (map.get("provider") instanceof String) {
			builder.provider((String) map.get("provider"));
		}
		Object deviceIdValue = map.get("device_id");
		if (deviceIdValue instanceof Number) {
			builder.deviceId(((Number) deviceIdValue).intValue());
		}
		return builder.build();
	}
}

File: kreuzberg/packages/php/src/Config/AccelerationConfig.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Config;

/**
 * Hardware acceleration configuration for ONNX Runtime models.
 *
 * Configures which execution provider to use for ONNX model inference,
 * enabling hardware acceleration on GPU devices (CUDA, TensorRT, CoreML).
 *
 * @example
 * ```php
 * use Kreuzberg\Config\AccelerationConfig;
 *
 * $acceleration = new AccelerationConfig(
 *     provider: 'cuda',
 *     deviceId: 0,
 * );
 * ```
 */
readonly class AccelerationConfig
{
    /**
     * @param string $provider Execution provider: "auto" (default), "cpu", "coreml", "cuda", "tensorrt"
     * @param int $deviceId GPU device ID (for CUDA/TensorRT). Default is 0.
     */
    public function __construct(
        public string $provider = 'auto',
        public int $deviceId = 0,
    ) {
    }

    /**
     * Create configuration from array data.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $provider = $data['provider'] ?? 'auto';
        $deviceId = $data['device_id'] ?? 0;

        return new self(
            provider: is_string($provider) ? $provider : 'auto',
            deviceId: is_int($deviceId) ? $deviceId : 0,
        );
    }

    /**
     * Convert configuration to array for FFI.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [];

        // Only add non-default values
        if ($this->provider !== 'auto') {
            $result['provider'] = $this->provider;
        }
        if ($this->deviceId !== 0) {
            $result['device_id'] = $this->deviceId;
        }

        return $result;
    }
}
